Implement a simple stateless observation executable

abailly-iohk opened the pull request:
:clipboard: Created a `hydra-chain-observer` executable, which prints a few basic observations of the head life-cycle.

:clipboard: The chain observer does connect directly to a `cardano-node` just the same as a ydra-node` would.

:clipboard: Refactored low-level observation functions to only take `UTxO` and `Tx` (especially the `observeCommitTx`).

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

  - One TODO in [`State.hs`](https://github.com/input-output-hk/hydra/blob/c8c9a7c25cbd4eff618d2609a932bea85f0f7b1c/hydra-node/src/Hydra/Chain/Direct/State.hs#L798) about a stateful generator - this will be removed once we get rid of stateful observation anyways

ch1bo committed:
Draft a ChainObserverSpec

This asserts the initialization of a Hydra head is also seen by the (not
yet existing) hydra-chain-observer executable

locallycompact committed:
hydra-chain-observer package

locallycompact committed:
Call hydra-chain-observer from ChainObserverSpec

ch1bo committed:
Provide node socket and network id to hydra-chain-observer

locallycompact committed:
Command line parsing of hydra-chain-observer

ch1bo committed:
Connect to a local node from hydra-chain-observer

locallycompact committed:
WIP

locallycompact committed:
ChainObserverSpec waits for matching output

ch1bo committed:
Produce a static HeadInitTx message on hydra-chain-observer

ch1bo committed:
Actually connect to cardano-node and print using tracer

ch1bo committed:
Do not use show on File to avoid quotes in file paths

locallycompact committed:
Fetch all blocks in hydra-chain-observer and print txs

ch1bo committed:
Call observeRawInitTx from chain observer

locallycompact committed:
WIP

locallycompact committed:
Check we can observe commit tx

abailly committed:
Introduce generic head tx observation

ch1bo committed:
Duplicate raw commit observation

ch1bo committed:
Add note about observing collect tx without looking at inputs

We might want to implement that strategy as an observeRawCollectTx
first and see whether that aligns with the "full" observation.

abailly committed:
Missing deriving strategies

abailly committed:
Add test to observe CollectCom Tx

abailly committed:
Use utxo to observe head txs

locallycompact committed:
Add observeHeadTx test

ch1bo committed:
Add more cases to observeHeadTx and its test

ch1bo committed:
Add abort txs to genStateWithTx and add to observeHeadTx

abailly committed:
Cover more of head lifecycle observation

locallycompact committed:
ChainObserver: add RollForward trace

ch1bo committed:
Add a FIXME where we need to fix our UTxO

abailly committed:
Correctly adjust UTxO when observing a Tx

locallycompact committed:
Add HeadAbortTx and HeadContestTx tracers

locallycompact committed:
Add hydra-chain-observer test suite

locallycompact committed:
hydra-chain-observer: add Test.ChainObserverSpec

abailly committed:
Check observer covers all kind of transactions

abailly committed:
Formatting

abailly committed:
Only update observer's UTxO upon observing a Head Tx

locallycompact committed:
clean up observeTx

ch1bo committed:
Add some FIXME notes to observeCommitTx before refactoring it

We plan to use the observeCommitTx also from the chain observer. To make
that happen we need to remove stateful information like the known
initials of a head.

However, those are (in a contrived way) currently our only way of
ensuring that this transaction is a proper Hydra Head transaction.

The FIXME describes alternatives how to achieve this via keeping the
seed in the datum or providing the full resolved inputs of a
transaction (including the committed UTxO).

locallycompact committed:
Change RawCommitObservation to CommitObservation

ch1bo committed:
Check commits are spending from vInitial in observeCommitTx

This ensures that the commit tx is part of the Head protocol (because
the validators must have ran).

locallycompact committed:
Create a test about observing commit transactions

This ensures we do not observe transactions which look like commit
transactions on their outputs, but do not spend from initial script.

ch1bo committed:
Relativize FIXMEs

abailly committed:
Add option to start observing the chain from some point

locallycompact committed:
nix/hydra/packages: add hydra-chain-observer

locallycompact committed:
query the tip when no startChainFrom given

ch1bo committed:
Find intersection with selected starting point

ch1bo committed:
Make warnings on hydra-chain-observer less pedantic

abailly committed:
Fix compilation errors

ch1bo committed:
Refactor EndToEndLog and create a smaller HydraNodeLog type

This avoids passing the "too big" type into the HydraClient

ch1bo committed:
Update CHANGELOG and hydra-chain-observer README

ch1bo committed:
Remove hydra-chain-observer/CHANGELOG

This package is released with all other hydra-* packages and shares a changelog

ch1bo commented on hydra-chain-observer/test/Hydra/ChainObserverSpec.hs:
```diff
@@ -0,0 +1,40 @@
+
+    prop "Does not updates UTxO state given transactions outside of Head lifecycle" $
+      forAll genSequenceOfSimplePaymentTransactions $ \(utxo, txs) ->
+        fst (observeAll testNetworkId utxo txs) === utxo
```
@abailly-iohk Do we need this? It's no harm if the UTxO is always updated, even when nothing is observed. I think it would simplify the implementation a bit if we remove this constraint.

ch1bo replied:
It's an optimization. That is, we do not accumulate memory on unrelated transactions.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -610,8 +611,8 @@ observeCommit ::
-  let initials = fst3 <$> initialInitials
-  observation <- observeCommitTx networkId initials tx
+  let utxo = getKnownUTxO st
+  observation <- observeCommitTx networkId utxo tx
```
SB: Shouldn't we make sure we only process commits of "our head" now?

Yes. This is implicit here (as only UTxO of our head are resolvable), but could be improved by an explicit `guard $ headId == commitHeadId` or so.

ch1bo replied:
Done

ch1bo committed:
Fix formatting

ch1bo committed:
Remove an unused BlockHeader match

ch1bo committed:
Remove embedRevision on non-versioned binary hydra-chain-observer

ch1bo committed:
Add explicit guard in observeCommit

This makes it clear that the hydra-node still only "cares" about commit
transactions about the head it knows in the chain layer.

ch1bo committed:
Add hydra-chain-observer to hydra-cluster test environment

v0d1ch reviewed (approved):
Good work guys!

ch1bo committed:
Trace known scripts in hydra-chain-observer

ch1bo commented:
Tested this manually against `preprod` to find the currently open head we use in our tactical meetings:

```shell
cabal run hydra-chain-observer -- --node-socket testnets/preprod/node.socket --testnet-magic 1 --start-chain-from "43251332.9ea784cd95687e3c90a17b6f3633b5554f25a3ceba82f860c2e4d3cb57a0b236" | jq 'select(.message.tag == "HeadInitTx")'
```

Will output:

```json
{
  "timestamp": "2023-11-13T14:17:57.570680734Z",
  "threadId": 6,
  "namespace": "hydra-chain-observer",
  "message": {
    "headId": "1185347034540e3a9c60c69a73185ebdcd3b06cd243047aea289e6b6",
    "tag": "HeadInitTx"
  }
}
```
